### ISSUE TITLE
feat(core): normalization per spec for transforms/etc 🙀 

### ISSUE THREAD

**Issue by srl295** [keymanapp/keyman#9468]
[CLDR-16943](https://unicode-org.atlassian.net/browse/CLDR-16943) details (or will detail) SC consensus about the role of Unicode normalization. Implement it.

**Comment by srl295** [comment]
- Cached Context needs to stay in NFD
- Engine needs to be able to do a normalization-insensitive compare
- Normalization should be available to KMN but not switched on now - KeyboardProcessor

**Comment by mcdurdin** [comment]
NFD marker tricks:

1. `U+0061 \m{marker1} U+0300` gives NFC of `à`. So what happens to the marker in the cached context? Is cached context NFD but app context is NFC? How do we sync?
2. `U+0061 \m{marker1} U+0300 \m{marker2} U+0320` goes to NFD `U+0061 U+0320 U+0300` which breaks markers, because we no longer know where they go. Should this be a compiler error? Or is there some clever way we can work around this (e.g. normalize both app context and cached context before comparison for equality?)

**Comment by srl295** [comment]
upstream CLDR normalization ticket was merged, but basically, we don't need the ticket, we need the behavior. So this is shovel ready.

**Comment by srl295** [comment]
So, I'm kind of thinking at this moment about *not* trying to normalize in kmc at all. The reason is, because the core side will already need to be able to normalize not just all strings in the compiled data, but  also the context. Secondly, it gets us out of having to even consider what version of node (or browser!) kmc is running under. This could even lead to a class of non-determinism in the compiler, where two runs of kmc give different kmx depending on the node version.  By a 'leave it alone' approach, we just write into kmx exactly whatever is in the xml.

**Comment by mcdurdin** [comment]
Hmm. I hear you on non-determinism. It should only impact unencoded scripts though, right? Given stability rules?

But... how will you do regex matching? It seems to me it would be difficult to normalize regexes post-construction.

**Comment by srl295** [comment]
> Hmm. I hear you on non-determinism. It should only impact unencoded scripts though, right? Given stability rules?

previously unencoded, yes. 

> But... how will you do regex matching? It seems to me it would be difficult to normalize regexes post-construction.

The regex pattern is just a string and can be normalized before constructing the matcher, I'd think?  (He says, confidently)

**Comment by mcdurdin** [comment]
> The regex pattern is just a string and can be normalized before constructing the matcher, I'd think? (He says, confidently)

`\uxxxx` says otherwise.

**Comment by mcdurdin** [comment]
And even worse, `[a-z]`?

**Comment by mcdurdin** [comment]
Ref https://unicode.org/reports/tr18/#Canonical_Equivalents

**Comment by srl295** [comment]
> > The regex pattern is just a string and can be normalized before constructing the matcher, I'd think? (He says, confidently)
> 
> `\uxxxx` says otherwise.

- `\u{xxxx}` is already processed by KMC. 
- So `\u{00E9}` will already be `E9 00` in UTF-16 in .kmx
- when core pulls it in, it can be normalized

if `\u{xxxx}` was processed later by the regex engine, then yes.  but that should only be necessary for syntactical elements.

**Comment by srl295** [comment]
> `[a-z]`

may need to parse and process such a range.

> _For example, the pattern should contain no characters that would not occur in that normalization form, nor sequences that would not occur._

characters can be checked for perhaps… sequences may be more challenging.

**Comment by mcdurdin** [comment]
Precisely. If I have a transform from `[\u{00E8}-\u{00EB}]` (èéêë) to `[\u{00EC}-\u{00EF}]` (ìíîï), how will that work with decomposition? Do we need to expand ranges (beware `0020-10FFFF`)?

**Comment by srl295** [comment]
> Precisely. If I have a transform from `[\u{00E8}-\u{00EB}]` (èéêë) to `[\u{00EC}-\u{00EF}]` (ìíîï), how will that work with decomposition? Do we need to expand ranges (beware `0020-10FFFF`)?

Could be a reason for limiting the size of ranges…if need be

We may end up from all of this needing to say, the regexes must be written in NFD and see the TR…

**Comment by srl295** [comment]
- maybe the spec 
-

**Comment by mcdurdin** [comment]
Let's discuss at our meeting tomorrow

**Comment by srl295** [comment]
Ok.  An issue with "push all normalization into core" is this … Identity.
`<key id="a" output="\u{00e0}" /><key id="b" output="\u{0061}\u0300}" />` will create two `strs` entries. 

At least it will mean that code can't use the `strs` index to _negatively_ test for string identity.

**Comment by srl295** [comment]
https://unicode-org.github.io/icu-docs/apidoc/released/icu4c/classNormalizer2.html for reference

**Comment by srl295** [comment]
From discussion, captured by @mcdurdin 
-----
Markers: 
1. `1 e\u{0300}\m{problem_marker}\u{0320}` (NFD) reorders `1 e\u{0320}\u{0300}` (without markers)
2. We then re-inject markers: 
   * `1 e\m{problem_marker}\u{0320}\u{0300}` (tie marker to following char), or 
   * `1 e\u{0320}\u{0300}\m{problem_marker}` (tie marker to preceding char), or
   * `1 e\m{problem_marker}\u{0320}\u{0300}` (insert marker as early as possible in norm. cluster)

Argument for following-char method is that it makes more sense for end-of-context:
* `1 e\u{0300}\u{0320}\m{problem_marker}` ==>  `1 e\u{0320}\u{0300}\m{problem_marker}` (tie marker to follow char, i.e. end-of-string)
* To implement this, work backwards from end-of-context to first stable char

Proposed Algorithm: 
1. First, we need to transform compat chars and NFC chars to NFD: break strings at markers, normalize each string to NFD
2. Concatenate the result, including markers, and re-break at stable characters
3. Re-normalize each chunk, removing markers and remembering attachment to following char, to get any NFD reordering, then re-inject markers

This algorithm can be applied to _any_ string, including: keyboard source strings, input context from app (cached context), and output in various stages from the processor (i.e. before and after each transform step).

----

What about repeated characters:
* `1 e\u{0300}\u{0320}\m{problem_marker}\u{0300}` --> `1 e\u{0320}\u{0300}\u{0300}` (where does marker go?) 

This hinges on: 
* Given two independent normalized-to-NFD strings, when concatenated, can we ever have codepoints change? (it is recognized that reorder is possible.) --> my expectation is that if this is the case, it will be one or two old sequences that we can special-case on, so we should be able to move forward.

Refs:
* https://unicode.org/reports/tr15/#Concatenation

---

Alternative solution: markers always move to end of a normalized sequence. 

`<key to="[\u{0320}\u{0300}]\m{problem_marker}">`

   * `<transform from="e\u{0320}\m{problem_marker}\u{0300}" to="HELLO">`
   * `<transform from="e\u{0320}\u{0300}\m{problem_marker}" to="GOODBYE">`

**Comment by srl295** [comment]
I'm kind of leaning towards marker-moves-to-end. 

It will make the marker make more sense with the NFC content. 

I think we could say that markers stay in the same order they were injected in.

**Comment by mcdurdin** [comment]
marker-moves-to-end of sequence: NFC seq or NFD seq? That is, will the marker remain interleaved with combining diacritics? Because if not, I think that's going to be troublesome for keyboard devs to figure out.

**Comment by srl295** [comment]
I think we landed on "tie marker to following char"
